File: 05_2_code_tue_raw_birth_death_IBM.py
# Continuous-time birth-death IBM

import numpy as np
import matplotlib.pyplot as plt

# Simulation control
rng = np.random.default_rng(95756)
t = 0
t_max = 30

# Initial biological state

# --individual scale
N_0 = 10
alive = np.full(N_0, 1)
b = np.full(N_0, 0.25)    # birth rate
m = np.full(N_0, 0.2)     # death rate

# --population scale
N_t = np.sum(alive) #scale transition (sum over individuals -> population)
N = N_t      #track N
times = t    #track time

# Simulate events
while t < t_max and N_t > 0:

    # calculate event intensity
    total_birth = np.sum(b * alive)
    total_death = np.sum(m * alive)
    intensity = total_birth + total_death

    # waiting time to next event
    w = rng.exponential(scale=1/intensity)

    # advance time
    t = t + w

    # which event type occurs?
    event_birth = rng.binomial(1, total_birth / intensity) #1 = birth

    # which individual does the event happen to?
    i = rng.choice(np.where(alive == 1)[0])  #equal p, identical individuals

    # update system state (do event)
    if event_birth:
        # add a new individual and its traits
        alive = np.append(alive, 1)
        b = np.append(b, 0.25)
        m = np.append(m, 0.2)
    else:
        # death
        alive[i] = 0

    N_t = sum(alive)
    N = np.append(N, N_t)
    times = np.append(times, t)

print(alive)

plt.figure()
plt.step(times, N, where='post')
plt.ylim(bottom=0)
plt.xlabel("Time")
plt.ylabel("N")
plt.title("Birth-death IBM")

File: 05_2_code_tue_raw_birth_death_IBM.R
# Continuous-time birth-death IBM

# Simulation control
set.seed(95756)
t <- 0
t_max <- 30

# Initial biological state

# --individual scale
N_0 <- 10
alive <- rep(1, N_0)
b <- rep(0.25, N_0)    # birth rate
m <- rep(0.2, N_0)     # death rate

# --population scale
N_t <- sum(alive)      # scale transition (sum over individuals -> population)
N <- N_t               # track N
times <- t             # track time

# Simulate events
while ( t < t_max & N_t > 0 ) {

    # calculate event intensity
    total_birth <- sum(b * alive)
    total_death <- sum(m * alive)
    intensity <- total_birth + total_death

    # waiting time to next event
    w <- rexp(1, rate=intensity)

    # advance time
    t <- t + w

    # which event type occurs?
    event_birth <- rbinom(1, size=1, prob=total_birth / intensity)

    # which individual does the event happen to?
    i <- sample(which(alive == 1), size=1)  #equal p, identical individuals

    # update system state (do event)
    if ( event_birth == 1 ) {

        # add a new individual and its traits
        alive <- c(alive, 1)
        b <- c(b, 0.25)
        m <- c(m, 0.2)

    } else {

        # death
        alive[i] <- 0

    }

    N_t <- sum(alive)
    N <- c(N, N_t)
    times <- c(times, t)
}

print(alive)

plot(times, N, type="s", ylim=c(0, max(N)), col="blue", xlab="Time", ylab="N",
    main="Birth-death IBM")
